ecommerce/models.py

```diff
@@ -154,3 +160,93 @@ def save(self, *args, **kwargs):
+
+    AMOUNT_TYPES = [PERCENT_DISCOUNT, FIXED_DISCOUNT]
+
+    # The coupon code used for redemption by the purchaser in the user interface.
+    # If blank, the purchaser may not redeem this coupon through the user interface,
+    # though it may be redeemed in their name by an administrator.
+    coupon_code = TextField(null=True, blank=True)
+    # One and only one of these three foreign keys must be set to not null
+    course_run = ForeignKey(CourseRun, on_delete=SET_NULL, null=True)
+    course = ForeignKey(Course, on_delete=SET_NULL, null=True)
+    program = ForeignKey(Program, on_delete=SET_NULL, null=True)
+
+    # percent or fixed discount
+    amount_type = CharField(
+        choices=[(_type, _type) for _type in AMOUNT_TYPES],
+        max_length=30,
+    )
+    # Either a number from 0 to 1 representing a percent, or a fixed value for discount
+    amount = DecimalField(decimal_places=2, max_digits=20)
+
+    # Number of people this coupon can be redeemed by
+    num_coupons_available = IntegerField(null=False)
+    # Number of times a person can redeem a coupon
+    num_redemptions = IntegerField(null=False)
+    # After this time the coupons will not be redeemable
+    expiration_date = DateTimeField(null=True)
+    # If true, coupons are not presently redeemable
+    disabled = BooleanField(default=False)
+
+    def __str__(self):
+        """Description for Coupon"""
+        product = self.course_run or self.course or self.program
+        return "Coupon {amount_type} {amount} for {product}, {num_coupons_available} left".format(
+            amount_type=self.amount_type,
+            amount=self.amount,
+            product=product,
+            num_coupons_available=self.num_coupons_available,
+        )
+
+
+class UserCoupon(Model):
+    """
+    Model for uses of a coupon by a user.
+    """
+    coupon = ForeignKey(Coupon, on_delete=SET_NULL, null=True)
+    user = ForeignKey(settings.AUTH_USER_MODEL, on_delete=SET_NULL, null=True)
+
+    # Number of times left the user can use the coupon in a purchase
+    available_redemptions = IntegerField()
+
+    class Meta:
+        unique_together = ('coupon', 'user',)
+
+    def __str__(self):
+        """Description for UserCoupon"""
+        return "UserCoupon for user {username} and coupon {coupon}, {available_redemptions} redemptions left".format(
+            username=self.user.username,
+            coupon=self.coupon,
+            available_redemptions=self.available_redemptions,
+        )
+
+
+class RedeemedCoupon(Model):
+    """
+    Model for coupon which has been used in a purchase by a user.
+    """
+    order = ForeignKey(Order, on_delete=SET_NULL, null=True)
+    coupon = ForeignKey(Coupon, on_delete=SET_NULL, null=True)
+
+    class Meta:
+        unique_together = ('order', 'coupon',)
+
+    def __str__(self):
+        """Description for RedeemedCoupon"""
+        return "RedeemedCoupon for order {order}, coupon {coupon}".format(
```

Comment:
If the `__str__()` method on `Coupon` contains the word "coupon", you don't need to _also_ include the word "coupon" here.